### PR TITLE
[TerritorialCouncil][CoPol][Admin] Add mayor and leader membership

### PR DESCRIPTION
**Pull request by illusionOfParadise**

### CODE REVIEW

**Comment by illusionOfParadise** [comment]
[ch12051]

File: src/Entity/TerritorialCouncil/PoliticalCommitteeMembership.php [hunk 1]
```diff
@@ -140,9 +141,28 @@ public function clearQualities(): void
 
     public function hasQuality(string $name): bool
     {
-        return $this->getQualities()->filter(function (PoliticalCommitteeQuality $quality) use ($name) {
-            return $quality->getName() === $name;
-        })->count() > 0;
+        if (0 === $this->qualities->count()) {
```
Comment (ottaviano) [inline src/Entity/TerritorialCouncil/PoliticalCommitteeMembership.php:144]: this method can use the other one:

```php
public function hasQuality(string $name): bool
{	   
    return $this->hasOneOfQualities([$name]);
}
```

File: src/Repository/ElectedRepresentative/MandateRepository.php [hunk 2]
```diff
@@ -74,4 +75,24 @@ public function findByFunctionAndUserListDefinitionForAdherent(
             ->getResult()
         ;
     }
+
+    public function hasMayorMandate(Adherent $adherent): bool
+    {
+        $mandates = $this
+            ->createQueryBuilder('mandate')
+            ->leftJoin('mandate.electedRepresentative', 'electedRepresentative')
```
Comment (ottaviano) [inline src/Repository/ElectedRepresentative/MandateRepository.php:83]: `innerJoin`

File: src/Repository/ElectedRepresentative/MandateRepository.php [hunk 3]
```diff
@@ -74,4 +75,24 @@ public function findByFunctionAndUserListDefinitionForAdherent(
...
+            ->leftJoin('mandate.politicalFunctions', 'politicalFunction')
```
Comment (ottaviano) [inline src/Repository/ElectedRepresentative/MandateRepository.php:84]: same: innerJoin

File: src/Repository/ElectedRepresentative/MandateRepository.php [hunk 4]
```diff
@@ -74,4 +75,24 @@ public function findByFunctionAndUserListDefinitionForAdherent(
...
+            ->andWhere('mandate.isElected = 1')
+            ->andWhere('mandate.onGoing = 1')
+            ->andWhere('politicalFunction.onGoing = 1')
+            ->setParameter('adherent', $adherent)
+            ->setParameter('mayor', TerritorialCouncilQualityEnum::MAYOR)
+            ->getQuery()
+            ->getResult()
+        ;
+
+        return \count($mandates) > 0;
```
Comment (ottaviano) [inline src/Repository/ElectedRepresentative/MandateRepository.php:96]: it's better to do a `SELECT COUNT(1)` in this case and avoid the doctrine hydration process:

```php
return 0 < (int) $this->createQueryBuilder()
  ->select(`COUNT(1)`)
  ...
  ->getSingleScalarResult()
```

File: src/Repository/TerritorialCouncil/PoliticalCommitteeMembershipRepository.php [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    public function __construct(RegistryInterface $registry)
+    {
+        parent::__construct($registry, PoliticalCommitteeMembership::class);
+    }
+
+    public function countLeaderAndMayorMembersFor(PoliticalCommittee $politicalCommittee): int
+    {
+        return (int) $this->createQueryBuilder('membership')
+            ->select('COUNT(1)')
+            ->leftJoin('membership.qualities', 'quality')
```
Comment (ottaviano) [inline src/Repository/TerritorialCouncil/PoliticalCommitteeMembershipRepository.php:22]: innerJoin

File: src/TerritorialCouncil/PoliticalCommitteeManager.php [hunk 6]
```diff
@@ -133,6 +159,88 @@ public function handleTerritorialCouncilMembershipUpdate(
...
+
+        if (!$adherent->hasPoliticalCommitteeMembership()) {
+            $this->throwException(
+                'political_committee.membership.adherent_has_no_membership',
+                [
+                    '{{ email }}' => $adherent->getEmailAddress(),
+                    '{{ territorialCouncil }}' => $territorialCouncil->getNameCodes(),
+                ]
+            );
+        }
```
Comment (ottaviano) [inline src/TerritorialCouncil/PoliticalCommitteeManager.php:231]: it seems to be the same piece of code for validation, maybe we can move this in a private method?

File: tests/TerritorialCouncil/PoliticalCommitteeManagerTest.php [hunk 7]
```diff
@@ -42,7 +61,12 @@ public function testCreateMembershipSuccessfully(): void
         $politicalCommittee = new PoliticalCommittee('PC name', new TerritorialCouncil());
         $qualityName = TerritorialCouncilQualityEnum::SENATOR;
 
-        $politicalCommitteeManager = new PoliticalCommitteeManager($this->entityManager);
+        $politicalCommitteeManager = new PoliticalCommitteeManager(
+            $this->entityManager,
+            $this->mandateRepository,
+            $this->membershipRepository,
+            $this->translator
+        );
```
Comment (ottaviano) [inline tests/TerritorialCouncil/PoliticalCommitteeManagerTest.php:69]: what about moving these lines into `setUp` method?

File: src/TerritorialCouncil/PoliticalCommitteeManager.php (hunk 6 above)
Comment (illusionOfParadise) [reply src/TerritorialCouncil/PoliticalCommitteeManager.php]: Only a part with first `if` is the same.

**Review by ottaviano — APPROVED** [review]

**Pull request merged**